sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -651,6 +651,10 @@ object ColumnPruning extends Rule[LogicalPlan] {
     // Can't prune the columns on LeafNode
     case p @ Project(_, _: LeafNode) => p
 
+    case p @ NestedColumnPruning(prunedNestedAttrs, projectsOnPrunedNestedAttrs)
```

Comment:
How about defining this rule outside the existing ColumnPruning? I feel this addition makes ColumnPruning more complicated.

Reply:
Oh, I missed to answer here, @maropu . This should be here before [line 660](https://github.com/apache/spark/pull/23873/files#diff-a636a87d8843eeccca90140be91d4fafR660). Otherwise, the projection is already moved down as the whole attribute instead of a subset of fields.

Reply:
Yea, I also checked @dbtsai 's answer: https://github.com/apache/spark/pull/23542#discussion_r248095908.
It looks reasonable.